From 4f2ba7599e22d3ee62f73c6a39457edfe2960bf2 Mon Sep 17 00:00:00 2001
From: yangyun <yangyun50@huawei.com>
Date: Tue, 6 May 2025 09:26:38 +0800
Subject: [PATCH 12/20] fuse: use exclusive lock when FUSE_I_CACHE_IO_MODE is
 set

mainline inclusion
from mainline-v6.12-rc1
commit 2f3d8ff457982f4055fe8f7bf19d3821ba22c376
category: bugfix
bugzilla: https://gitee.com/src-openeuler/kernel/issues/IBHLU4

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=2f3d8ff457982f4055fe8f7bf19d3821ba22c376

--------------------------------

This may be a typo. The comment has said shared locks are
not allowed when this bit is set. If using shared lock, the
wait in `fuse_file_cached_io_open` may be forever.

Fixes: 205c1d802683 ("fuse: allow parallel dio writes with FUSE_DIRECT_IO_ALLOW_MMAP")
CC: stable@vger.kernel.org # v6.9
Signed-off-by: yangyun <yangyun50@huawei.com>
Reviewed-by: Bernd Schubert <bschubert@ddn.com>
Signed-off-by: Miklos Szeredi <mszeredi@redhat.com>
Signed-off-by: Yifan Qiao <qiaoyifan4@huawei.com>
---
 fs/fuse/file.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/fs/fuse/file.c b/fs/fuse/file.c
index ea13f544c5..25d6951ef2 100644
--- a/fs/fuse/file.c
+++ b/fs/fuse/file.c
@@ -1355,7 +1355,7 @@ static bool fuse_dio_wr_exclusive_lock(struct kiocb *iocb, struct iov_iter *from
 
 	/* shared locks are not allowed with parallel page cache IO */
 	if (test_bit(FUSE_I_CACHE_IO_MODE, &fi->state))
-		return false;
+		return true;
 
 	/* Parallel dio beyond EOF is not supported, at least for now. */
 	if (fuse_io_past_eof(iocb, from))
-- 
2.54.0

